File: backend/app/models/marketplace_order.py
from __future__ import annotations

from uuid import UUID, uuid4

from sqlalchemy import ForeignKey, Integer, String, Uuid
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.models.mixins import TimestampMixin


class MarketplaceOrder(TimestampMixin, Base):
    """Commande boutique démo. Aucun Mobile Money réel n’est débité."""

    __tablename__ = "marketplace_orders"

    id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    user_id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True
    )
    product_id: Mapped[str] = mapped_column(String(64), nullable=False)
    product_name: Mapped[str] = mapped_column(String(120), nullable=False)
    amount: Mapped[int] = mapped_column(Integer, nullable=False)
    currency: Mapped[str] = mapped_column(String(8), nullable=False, default="XAF")
    status: Mapped[str] = mapped_column(String(16), nullable=False, default="recorded")
    note: Mapped[str | None] = mapped_column(String(255), nullable=True)

    user: Mapped[User] = relationship("User", back_populates="marketplace_orders")

File: lib/domain/entities/marketplace_order.dart
/// Commande boutique persistée (serveur + local). Paiement démo uniquement.
class MarketplaceOrder {
  const MarketplaceOrder({
    required this.productId,
    required this.productName,
    required this.amount,
    this.id,
    this.currency = 'XAF',
    this.status = 'recorded',
    this.message,
  });

  final String? id;
  final String productId;
  final String productName;
  final int amount;
  final String currency;
  final String status;
  final String? message;

  bool get isRecorded => status == 'recorded' || status == 'active';

  factory MarketplaceOrder.fromJson(Map<String, dynamic> json) {
    return MarketplaceOrder(
      id: json['id']?.toString(),
      productId: json['product_id'] as String? ?? '',
      productName: json['product_name'] as String? ?? '',
      amount: json['amount'] as int? ?? 0,
      currency: json['currency'] as String? ?? 'XAF',
      status: json['status'] as String? ?? 'recorded',
      message: json['message'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        if (id != null) 'id': id,
        'product_id': productId,
        'product_name': productName,
        'amount': amount,
        'currency': currency,
        'status': status,
        if (message != null) 'message': message,
      };
}
